NIAD-2539: Create Documentation for the mapping of resource type "Medication Statement". FHIR to HL7 and its reverse. (#15)
NHSDigital/patient-switching-adaptors-mapping-documentation: medication statements/README.md

# Medication Statement Mapping

## XML HL7 > JSON FHIR

A FHIR `MedicationStatement` is mapped from a XML HL7 `MedicationStatement` containing an `EhrSupplyAuthorise` component.</br>
When `ehrSupplyAuthorise` is referenced, it refers `MedicationStatement / component [@typeCode="COMP"] / ehrSupplyAuthorise`</br>
When `ehrSupplyDiscontinue` is referenced, it refers to find the first matching `ehrSupplyDiscontinue` where `ehrSupplyDiscontinue / reversalOf / priorMedicationRef / id [@root]` matching `MedicationStatement / component [@typeCode="COMP"] / ehrSupplyAuthorise / id [@root]` in the 'ehrExtract'</br>
When `ehrSupplyPrescribe` is referenced `MedicationStatement / component [@typeCode="COMP"] / ehrSupplyPrescribe`
When `EhrExtract` is referenced, it refers to the parent `EhrExtract` in the XML.</br>

> When `PracticeCode` is referenced, it refers to the losing practice ODS code.

Reviewer: I like this definitions block.

Reviewer: I added a bit of whitespace around the sentences, as I was finding it a bit hard to read.

# Medication Statement Mapping

## XML HL7 > JSON FHIR

A FHIR `MedicationStatement` is mapped from a XML HL7 `MedicationStatement` containing an `EhrSupplyAuthorise` component.

When `ehrSupplyAuthorise` is referenced, it refers to `MedicationStatement / component [@typeCode="COMP"] / ehrSupplyAuthorise`.

When `ehrSupplyDiscontinue` is referenced, it refers to find the first matching `ehrSupplyDiscontinue` where `ehrSupplyDiscontinue / reversalOf / priorMedicationRef / id [@root]` matching `MedicationStatement / component [@typeCode="COMP"] / ehrSupplyAuthorise / id [@root]` in the `EhrExtract`.

> When `ehrSupplyPrescribe` is referenced, it refers to `MedicationStatement / component [@typeCode="COMP"] / ehrSupplyPrescribe`.

Reviewer: It appears to do some filtering based on the `ehrSupplyAuthorise / id`

When `EhrExtract` is referenced, it refers to the parent `EhrExtract` in the XML.

When `PracticeCode` is referenced, it refers to the losing practice ODS code.

| Mapped to (JSON FHIR Medication Statement field) | Mapped from (XML HL7 / other source)                                                                                                                                       |
|--------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | id                                               | `ehrSupplyAuthorise / id [@root]` prefixed with `-MS`                                                                                                                      |

Reviewer: I think it's suffixed 😄

Author: corrected

| meta.profile\[0]                                 | fixed value = `"https://fhir.nhs.uk/STU3/StructureDefinition/CareConnect-GPC-MedicationStatement-1"`                                                                       |
| identifier.system                                | `https:\\PSSAdaptor\{{practiceCode}}` where {{practiceCode}} is losing practice ODS Code.                                                                                  |

> | identifier.value                                 | `ehrSupplyAuthorise / id [@root]` prefixed with `-MS`                                                                                                                      |

Reviewer: Suffixed.  I also think that markdown will want the [ to be escaped, but looks like it renders okay on GitHub, so not a biggie.

Author: Replaced Prefixed with Suffixed.  I think the [ doesn't need to be escaped as it's contain within a quotation block

| taken                                            | fixed value = `"unk"` <sup>1</sup>                                                                                                                                         |

> | basedOn.reference                                | a reference to a [MedicationRequest](../medication statements/README.md) using `ehrSupplyAuthorise / id [@root]` as the reference id                                       |

Reviewer: Hmm, this is referencing the same file 😕, which I'm now confused by.

| extension[0].url                                 | fixed value = `"https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-PrescribingAgency-1"`                                                               |
| extension[0].valueCodeableConcept.coding.system  | fixed value = `"https://fhir.nhs.uk/STU3/CodeSystem/CareConnect-PrescribingAgency-1"`                                                                                      |
| extension[0].valueCodeableConcept.coding.code    | fixed value = `"prescribed-at-gp-practice"`                                                                                                                                |
| extension[0].valueCodeableConcept.coding.display | fixed value = `"Prescribed at GP practice"`                                                                                                                                |

> | dosage[index].text                               | `MedicationStatement / pertinentInformation / pertinentMedicationDosage / text` <sup>2</sup>                                                                               |

Reviewer: Looks like there will only ever be 1 dosage (if I'm reading this code correctly), in which case `dosage[0].text` would be clearer.

Author: corrected

2. If there is no `pertinentInformation` which has a `pertinentInformationDosage / text`, then a default value of `"No Information available"` is used
3. This uses the most recent `availabilityTime` of an `ehrSupplyPrescribe` within any medication statement in the `ehrExtract` where there is an `inFulfilmentOf` which has a `priorMedication / id [@root]` matching the id of the current `MedicationStatement`<br/> If these are no matches then this extension is not added
4. This uses the values of `[@code]`, `[@displayName]` and `originalText` from `MedicationStatement / consumable / manufacturedProduct / manufacturedMaterial / code[0]` to create a unique key with a UUID value.  If this key has been used before then a `Medication` reference with the previously created UUID is used, otherwise a new UUID is generated. 
5. If there are no `consumable` with a `manufacturedProduct` with a `manufacturedMaterial` then this value is not set

> 6. The `ehrExtract` is parsed . If this has `ehrSupplyDiscontinue / availabilityTime / value` then value is set to `"STOPPED"` or `"COMPLETED"`

Reviewer: ```suggestion
6. The `ehrExtract` is parsed . If this has `ehrSupplyDiscontinue / availabilityTime / value` then value is set to `"STOPPED"` otherwise `"COMPLETED"`
```

7. If there is no matching `ehrSupplyDiscontinue` (see <sup>6</sup>) then if `ehrSupplyAuthorise / statusCode / code` = `"COMPLETE"` then `"COMPLETED"` is used, else `"ACTIVE"` is used

> 8. When `ehrSupplyDiscontinue / availabilityTime [@value]` does not exist and `status` = `"ACTIVE"` then this is set to the same value as `effectivePeriod.start`

Reviewer: Seems to fallback to `ehrSupplyAuthorise / effectiveTime / high \[@value]` and then falls back further to `medicationStatement / effectiveTime \[@value]`

Reviewer: Oh wait, I can see the Active behaviour now.  Looks like that's in addition to the fallbacks above.

Author: Yeah I agree, I've run through a number of scenarios and I can't see any circumstances where this logic will come into play, have removed this for clarity and added the fallback scenarios

9. When `ehrSupplyDiscontinue / availabilityTime [@value]` exists. If this exists but the referenced values do not then the availability time from either the `MedicationStatement`, `EhrExract` or `EhrComposition` is used. If none of these values exist then the value assigned to `effectivePeriod.end` is used instead

### Unmapped fields

> The following Medication Statement fields are not currently populated by the adaptor:

Reviewer: I think for the other resources we've mentioned only the fields which are described within GPConnect, but not populated by the PS Adaptor.

- informationSource
- note
- dosage.patientInstruction

Author: oh I didn't realise that, corrected as per the above

              "code": "prescribed-at-gp-practice",
              "display": "Prescribed at GP practice"
            }
          ]
        }
      }
    ],
    "identifier": [
      {
        "system": "https://PSSAdaptor/B83002",
        "value": "AEDDB53B-A30B-4EC9-A20F-A059C29A1C3E-MS"
      }
    ],
    "basedOn": [
      {
        "reference": "MedicationRequest/AEDDB53B-A30B-4EC9-A20F-A059C29A1C3E"
      }
    ],
    "status": "completed",
    "medicationReference": {
      "reference": "Medication/9fa0d37d-912f-44c2-855b-06d1e8623c97"
    },
    "effectivePeriod": {
      "start": "2010-02-10",
      "end": "2010-02-10"
    },
    "dateAsserted": "2010-02-10T08:31:19+00:00",
    "subject": {
      "reference": "Patient/c782c5ef-8235-4acf-945b-268a1bbe9d64"
    },
    "taken": "unk",
    "dosage": [
      {
        "text": "To Be Used As Directed"
      }
    ]
  }
}
```
</details>

## JSON FHIR > XML HL7

A XML HL7 `MedicationStatement` is mapped from a FHIR `medicationRequest`.  This can be one of two types, `ORDER` or `PLAN` which is determined by value of `medicationRequest.intent`

### Order

> For details of order mapping see the JSON FHIR > XML HL7 mapping for [Medication](../medications/README.md)

Reviewer: I probably don't know enough to know why the documentation for `MedicationStatement` would be within within `Medication` 🤔.

Author: XML MedicationStatements are mapped from from FHIR MedicationStatement, MedicationRequest and Medication.  They are all very much interlinked, but the medication mapping handles when an XML MedicationStatement is mapped from an "order".  I admit it can be confusing and might be best to move that mapping to here?

### Plan

| Mapped to (XML HL7)                                                                                                | Mapped from (JSON FHIR / other source )                                                                                            |
|--------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|

> | MedicationStatement \[@classCode] \[@moodCode]                                                                     | fixed value `@classCode="SBADM"`, fixed value `@moodCode="INT"`                                                                    |

Reviewer: I don't think we've used this syntax to combine two atttributes elsewhere, which may lead to confusion.

| MedicationStatement / statusCode                                                                                   | `medicationRequest.status` when `"active"` = `"ACTIVE"` otherwise is `"COMPLETE"`                                                  |
| MedicationStatement / effectiveTime / low \[@value]                                                                | `medicationRequest.dispenseRequest.validityPeriod.start`                                                                           |
| MedicationStatement / effectiveTime / high \[@value]                                                               | `medicationRequest.dispenseRequest.validityPeriod.end` <sup>1</sup>                                                                |
| MedicationStatement / availabilityTime  \[@value]                                                                  | `medicationRequest.dispenseRequest.validityPeriod.start`                                                                           |

> | MedicationStatement / consumable \[@typeCode] / manufacturedProduct \[@typeCode="CSM"]                             | fixed value = `"CSM"`                                                                                                              |

Reviewer: I find this confusing, `manufacturedProduct` doesn't have a typeCode from what I can see.

Author: yeah I think I've accidentally added an extra level on here by mistake

> | MedicationStatement / consumable / manufacturedProduct \[@typeCode="CSM"]                                          | fixed value = `"MANU"`                                                                                                             |

Reviewer: Should the `typeCode` here be `classCode` instead?

Author: Indeed it should

| MedicationStatement / pertinentInformation / pertinentMedicationDosage \[@classCode] \[@moodCode]                  | fixed value `@classCode="SBADM"`, fixed value `@moodCode="RMD"`                                                                    |
 | MedicationStatement / pertinentInformation / pertinentMedicationDosage / text                                      | `MedicationRequest.dosageInstruction[0].text`                                                                                      |
| MedicationStatement / participant \[@typeCode] \[@contextControlCode]                                              | fixed value = `@typeCode="AUT"`, fixed value = `@tcontextControlCode="OP"`. There may be zero, one or many participants            |
| MedicationStatement / participant / agentRef \[@classCode]                                                         | fixed value = `"AGNT"`                                                                                                             |

> | MedicationStatement / participant / agentRef / id \[@root]                                                         | See <sup>3</sup>                                                                                                                   |

Reviewer: ```suggestion
| MedicationStatement / participant / agentRef / id \[@root]                                                         | See <sup>2</sup>                                                                                                                   |
```

1. If `medicationRequest.dispenseRequest.validityPeriod.end` does not exist then this value is not populated

> 2. When `medicationRequest.requester.agent.referenceElement.resourceType` = `"Practitioner"` and `medicationRequest.requester.onBehalfOf` is populated then value is set from the reference to the `requester.agent` or `medicationRequest.requester.onBehalfOf`</br> Otherwise, when only `medicationRequest.requester.agent` exists then value is set from the reference to `requester.agent.reference` (when `resourceType` = `Practicioner`) or is the reference to the organisation (when `resourceType` = `Organisation`)</br>  If there are no matches for the previous options and `medicationRequest.recorder.reference` exists and the reference is either a `Practitioner`,`PractitionerRole` or `organisation` then the reference is set from the reference to `medicationRequest.recorder.reference` 

Reviewer: ```suggestion
2. When `medicationRequest.requester.agent.referenceElement.resourceType` = `"Practitioner"` and `medicationRequest.requester.onBehalfOf` is populated then value is set from the reference to the `medicationRequest.requester.agent` or `medicationRequest.requester.onBehalfOf`</br> Otherwise, when only `medicationRequest.requester.agent` exists then value is set from the reference to `requester.agent.reference` (when `resourceType` = `Practicioner`) or is the reference to the organisation (when `resourceType` = `Organisation`)</br>  If there are no matches for the previous options and `medicationRequest.recorder.reference` exists and the reference is either a `Practitioner`,`PractitionerRole` or `organisation` then the reference is set from the reference to `medicationRequest.recorder.reference` 
```

> ### Medication Code mapping

Reviewer: PR #13 adds a file `codeable concept/README.md`, I wonder if we could just reference that?

Author: Medication codeable concepts are slightly differently from other codeable concepts in the way that they are mapped as they do not contain extensions.

| ehrSupplyAuthorise / quantity / translation / originalText                       | `medicationRequest.dispenseRequest.quantity.unit`, `medicationRequest.dispenseRequest.extension[0].value`<sup>6</sup>, `medicationRequest.dispenseRequest.quantity.extension[0].value`<sup>7</sup><sup>8</sup>                                                                                                                                                                        |
| ehrSupplyAuthorise / pertinentInformation \[@typeCode]                           | fixed value = `"PERT"`                                                                                                                                                                                                                                                                                                                                                                |

> | ehrSupplyAuthorise / pertinentInformation / pertinentSupplyAnnotation / text     | `medicationRequest.dosageInstruction[0].patientInstruction` prefixed by `"Patient Instruction: "` & `medicationRequest.dispenseRequest.expectedSupplyDuration.value` prefixed by `"Expected Supply Duration"` and suffixed with `medicationRequest.dispenseRequest.expectedSupplyDuration.code` displayed as a unit of time & all `medicationRequest.note.text` prefixed by `"Notes"` |

Reviewer: ```suggestion
| ehrSupplyAuthorise / pertinentInformation / pertinentSupplyAnnotation / text     | `medicationRequest.dosageInstruction[0].patientInstruction` prefixed by `"Patient Instruction: "` & `medicationRequest.dispenseRequest.expectedSupplyDuration.value` prefixed by `"Expected Supply Duration: "` and suffixed with `medicationRequest.dispenseRequest.expectedSupplyDuration.code` displayed as a unit of time & all `medicationRequest.note.text` prefixed by `"Notes"` |
```

1. The `codableConcept` from `medicationStatement.extension[0].value` from the `ehrExtract` where `medicationStatement.basedOn` contains a reference to `medicationRequest.id` and `medicationStatement.extension[0].url` = `"https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-PrescribingAgency-1"` is used to generate the values.

> 2. When code.coding`.system` = `"https://fhir.nhs.uk/STU3/CodeSystem/CareConnect-PrescribingAgency-1"` and `code.coding.code` = `"prescribed-at-gp-practice"` or `"prescribed-by-previous-practice"` then `@code="394823007"` and `@displayName="NHS Prescription"`<br> When `code.coding.code` = `"prescribed-by-another-organisation"` then `@code="394828003"` and `@displayName="Prescription by another organisation"`. In both cases `@codeSystem="2.16.840.1.113883.2.1.3.2.4.15"`

Reviewer: ```suggestion
2. When `code.coding.system` = `"https://fhir.nhs.uk/STU3/CodeSystem/CareConnect-PrescribingAgency-1"` and `code.coding.code` = `"prescribed-at-gp-practice"` or `"prescribed-by-previous-practice"` then `@code="394823007"` and `@displayName="NHS Prescription"`<br> When `code.coding.code` = `"prescribed-by-another-organisation"` then `@code="394828003"` and `@displayName="Prescription by another organisation"`. In both cases `@codeSystem="2.16.840.1.113883.2.1.3.2.4.15"`
```

Reviewer: Although, actually I'm a bit confused by this statement in general.

Author: Have hopefully cleared this section up and explained it better

4. If `medicationRequest.dispenseRequest.validityPeriod.end` does not exist then this value is not populated
5. When `medicationRequest.extension.value.coding[0].code` = `"acute"` or `"acute-handwritten"` and `medicationRequest.extension.url` = `"https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-PrescriptionType-1"` then value used is `0`.</br> When `medicationRequest.extension.value.coding[0].code` = `"delayed-prescribing"`, `"repeat"` or `"repeat-dispensing"` then `medicationRequest.extension[0].value` when `medicationRequest.extension[0].url` = `"numberOfRepeatPrescriptionsAllowed"`.</br> If no value is found in either case then a default value of `1` is used.
6. When `medicationRequest.dispenseRequest.extension[0].url` = `"https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationQuantityText-1"`
7. When 'medicationRequest.dispenseRequest.quantity.extension[0].url' = `"https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationQuantityText-1"`

> 8. When no other values are found a default value of `"1"` is used.

Reviewer: ```suggestion
8. When no other values are found a default value of `"Unk UoM"` is used.
```

### EhrSupplyDiscontinue mapping

> This is part of the mapping for a `MedicationRequest` and is mapped when one of `MedicationRequest.extension.url` = `"https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationStatusReason-1"` (Medication Status Reason Stopped URL).  If this is not present then this mapping is omitted.

Reviewer: ```suggestion
This is part of the mapping for a `MedicationRequest` and is mapped when one of `MedicationRequest.extension\[index].url` = `"https://fhir.nhs.uk/STU3/StructureDefinition/Extension-CareConnect-GPC-MedicationStatusReason-1"` (Medication Status Reason Stopped URL).  If this is not present then this mapping is omitted.
```

| Mapped to (EhrSupplyDiscontinue XML HL7)                                                                           | Mapped from (JSON FHIR / other source )                                                                                                                                                                                                                 |
|--------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ehrSupplyDiscontinue / id \[@root]                                                                                 | adaptor generated UUID                                                                                                                                                                                                                                  |

> | ehrSupplyDiscontinue / code                                                                                        | set from the `codableConcept` in `medicationRequest.extension[0].extension[0].value`<sup>1</sup><sup>2</sup>                                                                                                                                            |

Reviewer: I think the double extension here adds confusion, perhaps add a definition to the top for the first extension for "MedicationStatusReason".

The second level of extension here, looks like it's also filtering on "statusReason".

Author: have done as suggested

            <pertinentSupplyAnnotation>
               <text>Expected Supply Duration: 150 day</text>
            </pertinentSupplyAnnotation>
         </pertinentInformation>
      </ehrSupplyPrescribe>
   </component>
   <pertinentInformation typeCode="PERT">
      <pertinentMedicationDosage classCode="SBADM" moodCode="RMD">
         <text>One To Be Taken At Night</text>
      </pertinentMedicationDosage>
   </pertinentInformation>
   <Participant typeCode="AUT" contextControlCode="OP">
      <agentRef classCode="AGNT">
         <id root="2D70F602-6BB1-47E0-B2EC-39912A59787D"/>
      </agentRef>
   </Participant>
</MedicationStatement>
```
</details>

### Plan

| Mapped to (XML HL7)                                                                                               | Mapped from (JSON FHIR / other source )                                                                                            |
|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|

> | MedicationStatement \[@classCode]                                                                                 | fixed value = `"SBADM"`                                                                                                            |

Reviewer 2: Some of these values are fixed in the HL7 MIM, so they would never be anything else. I brought this up with in the previous review of this table, some have been removed but there are still some present. I'll leave a comment against all the fixed ones (by the MIM).

| MedicationStatement / consumable / manufacturedProduct / manufacturedMaterial / code                              | Mapping for this code section detailed in the section below: [Medication Code Mapping](#medication-code-mapping)                   |
| MedicationStatement / component \[@typeCode="COMP"] / ehrSupplyAuthorise                                          | Mapping for `ehrSupplyAuthorise` is detailed in the section below: [EhrSupplyAuthorise Mapping](#ehrsupplyauthorise-mapping)       |
| MedicationStatement / component \[@typeCode="COMP"] / ehrSupplyDiscontinue \[@classCode="SPY"] \[@moodCode="PRQ"] | Mapping for `ehrSupplyDiscontinue` is detailed in the section below: [EhrSupplyDiscontinue Mapping](#ehrsupplydiscontinue-mapping) |
| MedicationStatement / pertinentInformation \[@typeCode]                                                           | fixed value = `"PERT"`                                                                                                             |

> | MedicationStatement / pertinentInformation / pertinentMedicationDosage \[@classCode]                              | fixed value = `"SBADM"`                                                                                                            |

Reviewer 2: "SBADM" is fixed by MIM, can remove
